app/services/catalog/update_order_item.rb

```diff
@@ -0,0 +1,58 @@
+module Catalog
+  class UpdateOrderItem
+    class OrderItemNotFound < StandardError; end
+    class ServiceInstanceNotFound < StandardError; end
+
+    def initialize(topic)
+      @payload = topic.payload
+      @message = topic.message
+    end
+
+    def process
+      Rails.logger.info("Processing service order topic message: #{@message} with payload: #{@payload}")
+
+      Rails.logger.info("Searching for OrderItem with a task_ref: #{@payload["task_id"]}")
+      order_item = find_order_item
+      Rails.logger.info("Found OrderItem: #{order_item.id}")
+
+      ManageIQ::API::Common::Request.with_request(order_item.context.transform_keys(&:to_sym)) do
+        order_item.update_message("info", "Task update message received with payload: #{@payload}")
+
+        if @payload["state"] == "completed"
+          order_item.state = "Order Completed"
+          order_item.update_message("info", "Order Complete")
+
+          external_url = fetch_external_url
+          order_item.external_url = external_url
+
+          Rails.logger.info("Updating OrderItem: #{order_item.id} with 'Order Completed' state and #{external_url}")
+          order_item.save!
+          Rails.logger.info("Finished updating OrderItem: #{order_item.id} with 'Order Completed' state")
+        end
+      end
+    end
+
+    private
+
+    def find_order_item
+      order_item = OrderItem.where(:topology_task_ref => @payload["task_id"]).first
```

Comment:
Could change this to `OrderItem.find_by!(:topology_task_ref => @payload["task_id"])`, and then rescue for `ActiveRecord::RecordNotFound`, that way you wouldn't have to define your own exception.

Reply:
Good call, I'm so used to not using `find_by` since rubocop used to whine about it that I blocked it out haha.